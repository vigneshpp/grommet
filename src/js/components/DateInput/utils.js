// Converting between Date and String types is handled via a "schema".
// The schema is an array of strings, split into strings with identical
// characters. So, 'mm/dd/yyyy' will be ['mm', '/', 'dd', '/', 'yyyyy'].

export const formatToSchema = format => {
  if (!format) return undefined;
  const result = [];

  let i = 0;
  let part;
  while (i < format.length) {
    if (!part || part[0] !== format[i]) {
      if (part) result.push(part);
      part = format[i];
    } else {
      part += format[i];
    }
    i += 1;
  }
  if (part) result.push(part);

  return result;
};

// convert value into text representation using the schema
export const valueToText = (value, schema) => {
  if (!value) return '';
  let text = '';

  const dates = (Array.isArray(value) ? value : [value]).map(v => new Date(v));

  let dateIndex = 0;
  let parts = {};
  schema.forEach(part => {
    const char = part[0].toLowerCase();
    // advance dateIndex if we already have this part
    while (
      dateIndex < dates.length &&
      (Number.isNaN(dates[dateIndex].date) ||
        ((char === 'm' || char === 'd' || char === 'y') && parts[char]))
    ) {
      dateIndex += 1;
      parts = {};
    }
    const date = dates[dateIndex];

    if (date && char === 'm') {
      text += date.getMonth() + 1;
      parts[char] = true;
    } else if (date && char === 'd') {
      text += date.getDate();
      parts[char] = true;
    } else if (date && char === 'y') {
      text += date.getFullYear();
      parts[char] = true;
    } else text += part;
  });

  return text;
};

const charCodeZero = '0'.charCodeAt(0);
const charCodeNine = '9'.charCodeAt(0);

const pullDigits = (text, index) => {
  let end = index;
  while (
    text.charCodeAt(end) >= charCodeZero &&
    text.charCodeAt(end) <= charCodeNine
  )
    end += 1;
  return text.slice(index, end);
};

export const textToValue = (text, schema) => {
  if (!text) return undefined;

  let result;
  const addDate = parts => {
    // do a little sanity checking on the values
    if (
      !parts.m ||
      !parts.d ||
      !parts.y ||
      parts.y.length < 4 ||
      parts.m.length > 2 ||
      parts.d.length > 2 ||
      parts.m > 12 ||
      parts.d > 31
    )
      return parts;
    const date = new Date(parts.y, parts.m - 1, parts.d).toISOString();
    if (!result) result = date;
    // single
    else if (Array.isArray(result)) result.push(date);
    // second
    else result = [result, date]; // third and beyond, unused?
    return {};
  };

  let parts = {};
  let index = 0;
  schema.forEach(part => {
    if (index < text.length) {
      const char = part[0].toLowerCase();
      if (parts[char] !== undefined) parts = addDate(parts);

      if (char === 'm') {
        parts.m = pullDigits(text, index);
        index += parts.m.length;
      } else if (char === 'd') {
        parts.d = pullDigits(text, index);
        index += parts.d.length;
      } else if (char === 'y') {
        parts.y = pullDigits(text, index);
        index += parts.y.length;
      } else if (text.slice(index, index + part.length) === part) {
        index += part.length;
      } else {
        // syntax error
        index = text.length;
        result = undefined;
      }
    }
  });
  addDate(parts);

  return result;
};
