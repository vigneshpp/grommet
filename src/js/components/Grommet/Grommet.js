import React, { Component } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ResponsiveContext,
  ThemeContext,
  ContainerTargetContext,
} from '../../contexts';

import {
  backgroundIsDark,
  deepMerge,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

class Grommet extends Component {
  static displayName = 'Grommet';

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      background: backgroundProp,
      dir,
      theme = {},
      themeMode,
    } = nextProps;
    const { theme: stateTheme, themeProp, themeModeProp } = prevState;

    const nextTheme = deepMerge(baseTheme, theme);
    if (!stateTheme || theme !== themeProp || themeMode !== themeModeProp) {
      const {
        colors: { background: themeBackground },
      } = nextTheme.global;

      // get initial value for dark so we can normalize background color
      nextTheme.dark = (themeMode || theme.defaultMode) === 'dark';
      const color = normalizeColor(
        backgroundProp || themeBackground,
        nextTheme,
      );

      // After normalizing, we set nextTheme.dark once more.
      // It is necessary that we set it twice. We have to handle two cases:
      // 1. Caller passes in a color object or a color name that resolves an
      //    object. In this case, we want to set dark as line 38 shows. The
      //    second set, in line 46, is a no-op.
      // 2. Caller passes a specific color value or a color name that resolves
      //    to a specific color value. In this case, we want dark to be set
      //    based on that color, which line 46 will do.
      // The double set of nextTheme.dark allows us to handle both cases here
      // without having to duplicate color object + name + dark mode detection
      // code here that is already in normalizeColor and backgroundIsDark.
      nextTheme.dark = backgroundIsDark(color, nextTheme);
      nextTheme.baseBackground = backgroundProp || themeBackground;

      if (dir) {
        nextTheme.dir = dir;
      }
      return {
        theme: nextTheme,
        themeProp: theme,
        themeModeProp: themeMode,
      };
    }

    return null;
  }

  state = {};

  componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onResize = () => {
    const { theme, responsive } = this.state;

    let breakpoint;
    // responsive would be undefined in the case of SSR or initial page load
    if (!responsive) {
      // In the case of SSR we'll need to use the user agent breakpoint
      breakpoint = this.deviceResponsive();
    }
    // Initial page load where both responsive and breakpoint are undefined
    if (!breakpoint) breakpoint = getBreakpoint(window.innerWidth, theme);

    if (breakpoint !== responsive) {
      this.setState({ responsive: breakpoint });
    }
  };

  deviceResponsive() {
    const { userAgent } = this.props;
    const { theme } = this.state;

    /*
     * Regexes provided for mobile and tablet detection are meant to replace
     * a full-featured specific library due to contributing a considerable size
     * into the bundle.
     *
     * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
     */
    if (userAgent) {
      if (
        /(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)
      ) {
        return getDeviceBreakpoint('tablet', theme);
      }
      if (/Mobile|iPhone|Android/.test(userAgent)) {
        return getDeviceBreakpoint('phone', theme);
      }
      return getDeviceBreakpoint('computer', theme);
    }
    return undefined;
  }

  render() {
    const {
      children,
      full,
      containerTarget = typeof document === 'object'
        ? document.body
        : undefined,
      ...rest
    } = this.props;
    delete rest.theme;
    const { theme, responsive: stateResponsive } = this.state;

    // Value from state should be correct once we resize
    // On first render we try to guess otherwise set the default as a tablet
    const responsive =
      stateResponsive ||
      this.deviceResponsive() ||
      theme.global.deviceBreakpoints.tablet;

    return (
      <ThemeContext.Provider value={theme}>
        <ResponsiveContext.Provider value={responsive}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <StyledGrommet full={full} {...rest}>
              {children}
            </StyledGrommet>
            {full && <FullGlobalStyle />}
          </ContainerTargetContext.Provider>
        </ResponsiveContext.Provider>
      </ThemeContext.Provider>
    );
  }
}

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
