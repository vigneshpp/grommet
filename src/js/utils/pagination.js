import { useCallback, useEffect, useState } from 'react';

export const normalizeShow = (showProp, step) => {
  let page;
  // by default, show refers to the index of an item,
  // but if using pagination, show can take the form of { page: # },
  // where page refers to the page # to show
  if (typeof showProp === 'number') page = Math.ceil((showProp + 1) / step);
  else if (typeof showProp === 'object' && 'page' in showProp)
    page = showProp.page;

  return page;
};

export const usePagination = ({ data, page, step, ...rest }) => {
  const totalPages = data ? Math.ceil(data.length / step) : 0;
  const [activePage, setActivePage] = useState(Math.min(page, totalPages) || 1);

  const itemsBeginIndex = step * (activePage - 1);
  const itemsEndIndex = itemsBeginIndex + step;
  const getCurrentItems = useCallback(
    items => {
      if (Array.isArray(items)) {
        return items.length
          ? items.slice(itemsBeginIndex, itemsEndIndex)
          : undefined;
      }
      return items;
    },
    [itemsBeginIndex, itemsEndIndex],
  );
  const [currentItems, setCurrentItems] = useState(getCurrentItems(data));

  useEffect(() => {
    if (data) {
      setCurrentItems(getCurrentItems(data));
    }
  }, [data, getCurrentItems, setCurrentItems]);

  const paginationProps = {
    numberItems: data && data.length,
    onChange: event => setActivePage(event.page),
    page,
    step,
    ...rest, // let anything coming from paginate prop override component
  };

  return [currentItems, paginationProps];
};
