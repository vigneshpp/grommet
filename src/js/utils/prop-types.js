import { PropTypes } from 'react-desc';

export const a11yTitlePropType = PropTypes.string.description(
  `Custom label to be used by screen readers. When provided, an aria-label will
   be added to the element.`,
);

export const colorPropType = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({ dark: PropTypes.string, light: PropTypes.string }),
]);

export const backgroundDoc = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    color: colorPropType,
    dark: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    image: PropTypes.string,
    position: PropTypes.string,
    opacity: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.oneOf(['weak', 'medium', 'strong']),
    ]),
    repeat: PropTypes.oneOfType([
      PropTypes.oneOf(['no-repeat', 'repeat']),
      PropTypes.string,
    ]),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['cover', 'contain']),
      PropTypes.string,
    ]),
    light: PropTypes.string,
  }),
]).description(`Either a color 
identifier to use for the background color. For example: 'neutral-1'. Or, a 
'url()' for an image. Dark is not needed if color is provided.`);

export const MARGIN_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
];

export const marginProp = PropTypes.oneOfType([
  PropTypes.oneOf(['none', ...MARGIN_SIZES]),
  PropTypes.shape({
    bottom: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    end: PropTypes.oneOfType([PropTypes.oneOf(MARGIN_SIZES), PropTypes.string]),
    horizontal: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    left: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    right: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    start: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    top: PropTypes.oneOfType([PropTypes.oneOf(MARGIN_SIZES), PropTypes.string]),
    vertical: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
  }),
  PropTypes.string,
]).description(`The amount of margin around the component. An object can
    be specified to distinguish horizontal margin, vertical margin, and
    margin on a particular side.`);

const PAD_SIZES = ['xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge'];

export const padPropType = PropTypes.oneOfType([
  PropTypes.oneOf(['none', ...PAD_SIZES]),
  PropTypes.shape({
    bottom: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    end: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    horizontal: PropTypes.oneOfType([
      PropTypes.oneOf(PAD_SIZES),
      PropTypes.string,
    ]),
    left: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    right: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    start: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    top: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    vertical: PropTypes.oneOfType([
      PropTypes.oneOf(PAD_SIZES),
      PropTypes.string,
    ]),
  }),
  PropTypes.string,
])
  .description(
    `The amount of padding around the box contents. An
    object can be specified to distinguish horizontal padding, vertical
    padding, and padding on a particular side of the box`,
  )
  .defaultValue('none');

export const genericProps = {
  a11yTitle: a11yTitlePropType,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch'])
    .description(`How to align along the cross axis when contained in
      a Box or along the column axis when contained in a Grid.`),
  gridArea: PropTypes.string.description(`The name of the area to place
    this inside a parent Grid.`),
  margin: marginProp,
};

export const hoverIndicatorPropType = PropTypes.oneOfType([
  PropTypes.bool,
  PropTypes.string,
  PropTypes.oneOf(['background']),
  PropTypes.shape({
    color: PropTypes.string,
    dark: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    image: PropTypes.string,
    light: PropTypes.string,
    position: PropTypes.string,
    opacity: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.oneOf(['weak', 'medium', 'strong']),
    ]),
    repeat: PropTypes.oneOfType([
      PropTypes.oneOf(['no-repeat', 'repeat']),
      PropTypes.string,
    ]),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['cover', 'contain']),
      PropTypes.string,
    ]),
  }),
]);

export const pointPropType = PropTypes.oneOf([
  'circle',
  'diamond',
  'square',
  'star',
  'triangle',
  'triangleDown',
]);
