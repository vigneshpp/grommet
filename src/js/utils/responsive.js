export const getBreakpoint = (windowWidth, theme) => {
  let result;
  const sortedBreakpoints = Object.keys(theme.global.breakpoints).sort(
    (a, b) => {
      const first = theme.global.breakpoints[a];
      const second = theme.global.breakpoints[b];
      if (!first) return 1;
      if (!second) return -1;
      if (!first.value) return 1;
      if (!second.value) return -1;
      return first.value - second.value;
    },
  );
  // the last breakpoint on the sorted array should have
  // no windowWidth boundaries
  const lastBreakpoint = sortedBreakpoints[sortedBreakpoints.length - 1];
  sortedBreakpoints.some(name => {
    const breakpoint = theme.global.breakpoints[name];
    if (breakpoint) {
      if (!breakpoint.value || breakpoint.value >= windowWidth) {
        result = name;
        return true;
      }
    }
    return false;
  });
  return result || lastBreakpoint;
};

export const getDeviceBreakpoint = (type, theme) =>
  theme.global.deviceBreakpoints[type];
