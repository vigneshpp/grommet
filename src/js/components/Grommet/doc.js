import { describe, PropTypes } from 'react-desc';

import { backgroundDoc } from '../../utils/prop-types';
import { getAvailableAtBadge } from '../../utils/mixins';

export const doc = Grommet => {
  const DocumentedGrommet = describe(Grommet)
    .availableAt(getAvailableAtBadge('Grommet', 'Utilities'))
    .description('The top level Grommet container.')
    .usage(
      `import { Grommet } from 'grommet';
<Grommet>...</Grommet>`,
    )
    .intrinsicElement('div');

  DocumentedGrommet.propTypes = {
    background: backgroundDoc,
    dir: PropTypes.oneOf(['rtl']).description(
      'Layout direction for right to left contexts',
    ),
    full: PropTypes.bool
      .description('Whether to take the whole viewport.')
      .defaultValue(false),
    plain: PropTypes.bool
      .description(
        `Whether or not Grommet should apply a global font-family, font-size,
        and line-height.`,
      )
      .defaultValue(false),
    cssVars: PropTypes.bool
      .description('Whether to expose the css variables.')
      .defaultValue(false),
    theme: PropTypes.object.description(
      'Custom styles for Grommet app component.',
    ),
    themeMode: PropTypes.oneOf(['dark', 'light']).description(
      `Dark vs. light theme variation. Default is unspecified and left to
      theme.`,
    ),
    userAgent: PropTypes.string.description(
      `User agent used to detect the device width for setting the initial
      breakpoint.`,
    ),
    containerTarget: PropTypes.object.description(
      `The node where Drop and Layer containers are inserted. Defaults to
      document.body which is almost always the right choice. This is used
      for less common cases like rendering within an internal node (e.g.
      shadow root).`,
    ),
  };

  return DocumentedGrommet;
};

export const themeDoc = {
  'grommet.extend': {
    description: 'Any additional style for Grommet.',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
  'global.font.face': {
    description: 'Custom font face declaration',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
};
