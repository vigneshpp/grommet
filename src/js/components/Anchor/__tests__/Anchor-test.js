import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';

import { axe } from 'jest-axe';
import { findAllByType } from '../../../utils';

import { Grommet } from '../../Grommet';
import { Anchor } from '..';

describe('Anchor', () => {
  afterEach(cleanup);

  test('should have no accessibility violations', async () => {
    const { container } = render(
      <Grommet>
        <Anchor>Link</Anchor>
      </Grommet>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  test('renders', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders with children', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor href="#">children</Anchor>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('warns about invalid label render', () => {
    console.warn = jest.fn();
    const warnSpy = jest.spyOn(console, 'warn');
    const { container } = render(
      <Grommet>
        <Anchor href="#" label="Test">
          invalid
        </Anchor>
      </Grommet>,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(warnSpy).toHaveBeenCalledWith(
      'Anchor should not have children if icon or label is provided',
    );

    warnSpy.mockReset();
    warnSpy.mockRestore();
    console.warn.mockReset();
  });

  test('warns about invalid icon render', () => {
    console.warn = jest.fn();
    const warnSpy = jest.spyOn(console, 'warn');
    const { container } = render(
      <Grommet>
        <Anchor href="#" icon={<svg />}>
          invalid
        </Anchor>
      </Grommet>,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(warnSpy).toHaveBeenCalledWith(
      'Anchor should not have children if icon or label is provided',
    );

    warnSpy.mockReset();
    warnSpy.mockRestore();
    console.warn.mockReset();
  });

  test('primary renders', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor href="#" primary label="Test" />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('focus renders', () => {
    const { container, getByText } = render(
      <Grommet>
        <Anchor href="#" label="Test" />
      </Grommet>,
    );
    fireEvent.focus(getByText('Test'));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('disabled renders', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor disabled />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('icon label renders', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor icon={<svg />} label="Test" onClick={() => {}} />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('reverse icon label renders', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor reverse icon={<svg />} label="Test" onClick={() => {}} />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('is clickable', () => {
    const onClick = jest.fn();
    const component = renderer.create(
      <Grommet>
        <Anchor href="#" label="Test" onClick={onClick} />
      </Grommet>,
    );
    const tree = component.toJSON();

    const anchor = findAllByType(tree, 'a');
    anchor[0].props.onClick();
    expect(onClick).toBeCalled();
  });

  test('renders tag', () => {
    const component = renderer.create(
      <Grommet>
        <Anchor href="#" label="Test" as="span" />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
