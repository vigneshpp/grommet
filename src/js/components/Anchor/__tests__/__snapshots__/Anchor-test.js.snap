// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor disabled renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  opacity: 0.3;
  cursor: default;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <a
    className="c1"
    disabled={true}
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`Anchor focus renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 > circle,
.c1 > ellipse,
.c1 > line,
.c1 > path,
.c1 > polygon,
.c1 > polyline,
.c1 > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c1::-moz-focus-inner {
  border: 0;
}

<div
  class="c0"
>
  <a
    class="c1"
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`Anchor icon label renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <a
    className="c1"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <span
      className="c2"
      style={
        Object {
          "display": "inline-flex",
        }
      }
    >
      <svg
        color="#7D4CDB"
      />
      <div
        className="c3"
      />
      Test
    </span>
  </a>
</div>
`;

exports[`Anchor primary renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <a
    className="c1"
    href="#"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    Test
  </a>
</div>
`;

exports[`Anchor renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <a
    className="c1"
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`Anchor renders tag 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <span
    className="c1"
    href="#"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    Test
  </span>
</div>
`;

exports[`Anchor renders with children 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <a
    className="c1"
    href="#"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    children
  </a>
</div>
`;

exports[`Anchor reverse icon label renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  className="c0"
>
  <a
    className="c1"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <span
      className="c2"
      style={
        Object {
          "display": "inline-flex",
        }
      }
    >
      Test
      <div
        className="c3"
      />
      <svg
        color="#7D4CDB"
      />
    </span>
  </a>
</div>
`;

exports[`Anchor should have no accessibility violations 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <div
    class="c0"
  >
    <a
      class="c1"
    >
      Link
    </a>
  </div>
</div>
`;

exports[`Anchor warns about invalid icon render 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 12px;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  class="c0"
>
  <a
    class="c1"
    href="#"
  >
    <svg
      color="#7D4CDB"
    />
  </a>
</div>
`;

exports[`Anchor warns about invalid label render 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  class="c0"
>
  <a
    class="c1"
    href="#"
  >
    Test
  </a>
</div>
`;
